core/blockchain.go

```diff
@@ -2263,41 +2263,55 @@ func (bc *BlockChain) ReadValidatorStats(
 
 // UpdateValidatorVotingPower writes the voting power for the committees
 func (bc *BlockChain) UpdateValidatorVotingPower(
-	batch rawdb.DatabaseWriter, state *shard.State,
+	batch rawdb.DatabaseWriter,
+	newEpochSuperCommittee, currentEpochSuperCommittee *shard.State,
 ) error {
-	if state == nil {
-		return errors.New("[UpdateValidatorVotingPower] Nil shard state")
+	if newEpochSuperCommittee == nil {
+		return shard.ErrSuperCommitteeNil
 	}
 
-	rosters := make([]votepower.RosterPerShard, len(state.Shards))
+	rosters := make([]*votepower.Roster, len(newEpochSuperCommittee.Shards))
 
-	for i := range state.Shards {
-		roster, err := votepower.Compute(state.Shards[i].Slots)
+	for i := range newEpochSuperCommittee.Shards {
+		subCommittee := &newEpochSuperCommittee.Shards[i]
+		roster, err := votepower.Compute(subCommittee)
 		if err != nil {
 			return err
 		}
-		rosters[i] = votepower.RosterPerShard{state.Shards[i].ShardID, roster}
+		rosters[i] = roster
 	}
 
 	networkWide := votepower.AggregateRosters(rosters)
-
 	for key, value := range networkWide {
-		statsFromDB, _ := rawdb.ReadValidatorStats(bc.db, key)
-		if statsFromDB == nil {
-			statsFromDB = &staking.ValidatorStats{
-				big.NewInt(0), numeric.NewDec(0),
-				[]staking.VotePerShard{}, []staking.KeysPerShard{},
-			}
+		stats, err := rawdb.ReadValidatorStats(bc.db, key)
+		if err != nil {
+			stats = staking.NewEmptyStats()
 		}
-		statsFromDB.TotalEffectiveStake = value.TotalEffectiveStake
-		statsFromDB.VotingPowerPerShard = value.VotingPower
-		statsFromDB.BLSKeyPerShard = value.BLSPublicKeysOwned
-		// TODO(audit): should clear the voting power of those validators who are not elected.
-		if err := rawdb.WriteValidatorStats(batch, key, statsFromDB); err != nil {
+		total := big.NewInt(0)
+		for i := range value {
+			total.Add(total, value[i].EffectiveStake)
+		}
+		stats.TotalEffectiveStake = total
+		stats.MetricsPerShard = value
+		if err := rawdb.WriteValidatorStats(
+			batch, key, stats,
+		); err != nil {
 			return err
 		}
 	}
 
+	existing, replacing :=
+		currentEpochSuperCommittee.StakedValidators(),
+		newEpochSuperCommittee.StakedValidators()
+	for currentValidator := range existing.LookupSet {
+		if _, keptSlot := replacing.LookupSet[currentValidator]; !keptSlot {
+			rawdb.DeleteValidatorSnapshot(
```

Comment:
why delete snapshot? They should be kept in db all the time.

Reply:
Because they aren’t in committee anymore so why hold onto that data

Reply:
Otherwise we will have an explosion of validator snapshots